AG-9015 - Add basic 'Hello World' implementation error bars

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-9015

This is the basic initial 'Hello World' implementation of error bars. This initial implementation is built with the intent of not relying on `CartesianSeries` too much to facilitate migration into the enterprise module later on.

This implementation has many limitations:

- A hardcoded placeholder styling is used.
- No caps are drawn (just the whiskers).
- Only the y-axis supports error bars.
- The y-axis domain is not extended to fit error bar data values. Values outside the domain are converted into `NaN` pixel coordinates and are therefore not drawn.
- All error bars are part of one group (no subgrouping).

olegat committed:
AG-9015 Add the LineSeries.errorBar property

olegat committed:
AG-9015 Define SeriesGroupZIndexSubOrderType

This is the type of the first parameter of getGroupZIndexSubOrder. This
type is being redefined in several places and might need an 'errorBars'
sub-group added. So preemptively refactor this by defining the type in
just one place.

olegat commented on packages/ag-charts-community/src/chart/series/cartesian/lineSeries.ts:
```diff
@@ -59,6 +59,26 @@ class LineSeriesLabel extends Label {
 type LineContext = SeriesNodeDataContext<LineNodeDatum>;
 type LineAnimationData = CartesianAnimationData<LineContext>;
 
+// TODO(olegat) move somewhere else?
```
Please advise whether these kind of `// TODO` comments are acceptable. If not, then I'll make a note of this somewhere else.

alantreadway replied:
We don't typically use these - there is a JIRA plugin which adds checklist support, so for smaller things like this I would usually capture them there, assuming you are going to resolve this in the scope of your current ticket?

olegat replied:
Understood. I'll remove this comment.

olegat committed:
Run prettier

olegat committed:
AG-9015 Initial 'Hello World' error bars implementation

olegat committed:
AG-9015 Fix circular dependency

alantreadway commented on packages/ag-charts-community/src/chart/series/cartesian/lineSeries.ts:
```diff
@@ -59,6 +59,26 @@ class LineSeriesLabel extends Label {
 type LineAnimationData = CartesianAnimationData<LineContext>;
 
+// TODO(olegat) move somewhere else?
+export class ErrorBarDatum {
```
Nit: We usually reserve `Datum` as a term for actual data - I would probably just run with `ErrorBar` or `ErrorBarConfig` perhaps?

olegat replied:
ACK. I'll use `ErrorBarConfig`

alantreadway commented on packages/ag-charts-community/src/chart/series/cartesian/lineSeries.ts:
```diff
@@ -59,6 +59,26 @@ class LineSeriesLabel extends Label {
+// TODO(olegat) move somewhere else?
+export class ErrorBarDatum {
+    @Validate(OPT_STRING)
+    yLowerKey: string;
```
A quirk/smell of TS->JS is that if a field is not assigned a value, it doesn't exist at runtime - typically you'll see that fields that are `@Validate()`'ed also assign an initial value, even if that is `undefined` - this allows us to sanity check that options provided by users actually exist before assignment from the users options, rather than just blinding adding.

olegat replied:
Thanks for the explanation. That makes sense. What values should be used?

I'm thinking of using an empty string `''` for `yLowerKey` and `yUpperKey` because these properties are required, and `undefined` for `yLowerName` and `yUpperName` because these are optional.

alantreadway commented on packages/ag-charts-community/src/chart/series/cartesian/lineSeries.ts:
```diff
@@ -118,6 +138,8 @@ export class LineSeries extends CartesianSeries<LineContext> {
     @Validate(OPT_STRING)
     yName?: string = undefined;
 
+    errorBar?: ErrorBarDatum = undefined;
```
You may need to assign an instance here, otherwise the options assignment code won't know how to construct it - once we extract to an enterprise module, the enterprise module intiialisation will take care of this step intrinsically though.

olegat replied:
If I assign this to a `new ErrorBarConfig()`, how will the `LineSeries` class know whether or not the `'errorBars'` property was given in the input JSON?

olegat committed:
AG-9015 Code review changes

olegat committed:
AG-9015 Remove hasErrorBars property

Instead, check if the errorBars property is defined. The errorBars
property needs to move from the LineSeries class into the
CartesianSeries base class for this to work.

This fixes a bug where an error is raised if the `errorBar` property is
not set in the input JSON.

olegat committed:
AG-9015 Simplify code by remove overrides

olegat committed:
AG-9015 Simplify code by removing errorBarUtil.ts

This file doesn't really do much. Move the implementation straight into
cartesianSeries.ts for now.

olegat committed:
AG-9015 Rename ErrorBarDatum to ErrorBarPoints

The ErrorBar class (Node subclass) doesn't actually need to know the data
values, it just needs to know the pixel coordinates to draw.

Also fix a bug in updateErrorBarNodes where nothing was being drawn
because the wrong input type was being provided.

olegat committed:
AG-9015 Rename ErrorBar to ErrorBarNode

olegat committed:
AG-9015 Move errorBar code into errorBar.ts

The implementation of error bars needs to be decoupled from the
CartesianSeries as much as possible.

The only thing that the error bar scene node need to know for now is the
the data points and X and Y scales (to convert data points into pixel
points). These variables are provided as parameters of the
ErrorBar.createNodeData method.

olegat committed:
Run prettier

olegat committed:
Revert "AG-9015 Define SeriesGroupZIndexSubOrderType"

This reverts commit b7c10902753342dfba793864c1262789e977b556.

olegat committed:
AG-9015 Remove inexistent import

alantreadway commented on packages/ag-charts-community/src/chart/errorBar.ts:
```diff
@@ -0,0 +1,125 @@
+export class ErrorBars {
+    public groupNode?: Group = undefined;
+    private selection?: Selection<ErrorBarNode> = undefined;
+    private nodeData: (ErrorBarPoints | undefined)[] = new Array(0);
```
Nit: `[]` is sufficient/convention for empty arrays.

olegat replied:
Done.

alantreadway commented on packages/ag-charts-community/src/chart/errorBar.ts:
```diff
@@ -0,0 +1,125 @@
+
+export class ErrorBars {
+    public groupNode?: Group = undefined;
+    private selection?: Selection<ErrorBarNode> = undefined;
```
Suggestion: If this field is initalised once in the constructor, you don't need to mark it as optional or assign an initial value inline here. You can also mark this field `readonly` if the reference never changes.

olegat replied:
Done. I've removed the `parent: Node` parameter of `createNodeData` and made it a parameter of the constructor.

alantreadway commented on packages/ag-charts-community/src/chart/errorBar.ts:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    update() {
+        if (this.selection != undefined) {
```
This guard should be redundant if you can apply my suggestion regarding the `selection` field above 🤞

olegat replied:
Done.

alantreadway reviewed (commented):
Some suggestions, but this looks like a reasonable first pass :-)

iMoses commented on packages/ag-charts-community/src/chart/errorBar.ts:
```diff
@@ -0,0 +1,125 @@
+        this.selection = undefined;
+
+        const { nodeData } = this;
+        const { yLowerKey = undefined, yUpperKey = undefined } = config ?? {};
```
```suggestion
        const { yLowerKey, yUpperKey } = config ?? {};
```
nit: it's redundant to assign `undefined` when spreading, that's the default anyhow

olegat replied:
Done.

olegat committed:
AG-9015 Initialise ErrorBar fields in the constructor.

This moves the `parent: Node` parameter from `createNodeData` to the
constructuor.

alantreadway reviewed (approved):
LGTM - thanks for making the updates!
